File: mono/sdks/python/src/rtvortex_sdk/streaming.py
"""SSE stream parser for RTVortex review streaming responses."""

from __future__ import annotations

from collections.abc import AsyncIterator, Iterator
from typing import TYPE_CHECKING

from rtvortex_sdk.models import ProgressEvent

if TYPE_CHECKING:
    import httpx


def _parse_sse_block(block: str) -> ProgressEvent | None:
    """Parse a single SSE text block into a ProgressEvent."""
    import json

    event_type = ""
    data_lines: list[str] = []

    for line in block.splitlines():
        if line.startswith("event:"):
            event_type = line[len("event:"):].strip()
        elif line.startswith("data:"):
            data_lines.append(line[len("data:"):].strip())

    if not data_lines:
        return None

    raw = "\n".join(data_lines)
    try:
        payload = json.loads(raw)
    except json.JSONDecodeError:
        payload = {"message": raw}

    if event_type:
        payload.setdefault("event", event_type)

    return ProgressEvent.model_validate(payload)


def iter_sse_events(response: httpx.Response) -> Iterator[ProgressEvent]:
    """Iterate over SSE events from a *synchronous* httpx streaming response.

    Usage::

        with client.stream("GET", url) as resp:
            for event in iter_sse_events(resp):
                print(event)
    """
    buf = ""
    for chunk in response.iter_text():
        buf += chunk
        while "\n\n" in buf:
            block, buf = buf.split("\n\n", 1)
            block = block.strip()
            if not block:
                continue
            evt = _parse_sse_block(block)
            if evt is not None:
                yield evt


async def aiter_sse_events(
    response: httpx.Response,
) -> AsyncIterator[ProgressEvent]:
    """Iterate over SSE events from an *async* httpx streaming response.

    Usage::

        async with client.stream("GET", url) as resp:
            async for event in aiter_sse_events(resp):
                print(event)
    """
    buf = ""
    async for chunk in response.aiter_text():
        buf += chunk
        while "\n\n" in buf:
            block, buf = buf.split("\n\n", 1)
            block = block.strip()
            if not block:
                continue
            evt = _parse_sse_block(block)
            if evt is not None:
                yield evt

File: mono/sdks/node/src/streaming.ts
/**
 * SSE streaming utilities for the RTVortex SDK.
 */

import type { ProgressEvent } from "./types.js";

/**
 * Parse a single SSE text block into a ProgressEvent.
 */
export function parseSSEBlock(block: string): ProgressEvent | null {
  let eventType = "";
  const dataLines: string[] = [];

  for (const line of block.split("\n")) {
    if (line.startsWith("event:")) {
      eventType = line.slice("event:".length).trim();
    } else if (line.startsWith("data:")) {
      dataLines.push(line.slice("data:".length).trim());
    }
  }

  if (dataLines.length === 0) return null;

  const raw = dataLines.join("\n");
  let payload: Record<string, unknown>;
  try {
    payload = JSON.parse(raw);
  } catch {
    payload = { message: raw };
  }

  if (eventType) {
    payload.event ??= eventType;
  }

  return payload as unknown as ProgressEvent;
}

/**
 * Async generator that yields ProgressEvents from a ReadableStream (SSE).
 *
 * Works with the native `fetch()` body stream in Node 18+ and browsers.
 */
export async function* iterSSEEvents(
  stream: ReadableStream<Uint8Array>,
): AsyncGenerator<ProgressEvent, void, undefined> {
  const decoder = new TextDecoder();
  const reader = stream.getReader();
  let buf = "";

  try {
    while (true) {
      const { done, value } = await reader.read();
      if (done) break;
      buf += decoder.decode(value, { stream: true });

      while (buf.includes("\n\n")) {
        const idx = buf.indexOf("\n\n");
        const block = buf.slice(0, idx).trim();
        buf = buf.slice(idx + 2);
        if (!block) continue;
        const evt = parseSSEBlock(block);
        if (evt) yield evt;
      }
    }

    // flush remainder
    const rest = buf.trim();
    if (rest) {
      const evt = parseSSEBlock(rest);
      if (evt) yield evt;
    }
  } finally {
    reader.releaseLock();
  }
}
